File: packages/i18nify-js/src/modules/banking/__tests__/validateIFSC.test.ts
import validateIFSC from '../validateIFSC';

describe('banking - validateIFSC', () => {
  describe('valid IFSCs', () => {
    it('should accept a valid IFSC (HDFC0001234)', () => {
      expect(validateIFSC('HDFC0001234')).toBe(true);
    });

    it('should accept a valid IFSC (SBIN0000001)', () => {
      expect(validateIFSC('SBIN0000001')).toBe(true);
    });

    it('should accept a branch code containing letters', () => {
      expect(validateIFSC('ICIC0ABC123')).toBe(true);
    });

    it('should accept a lowercase IFSC (case-insensitive)', () => {
      expect(validateIFSC('hdfc0001234')).toBe(true);
    });

    it('should accept an IFSC with surrounding whitespace', () => {
      expect(validateIFSC('  HDFC0001234  ')).toBe(true);
    });
  });

  describe('invalid IFSCs', () => {
    it("should reject an IFSC whose 5th char is not '0'", () => {
      expect(validateIFSC('HDFC1001234')).toBe(false);
    });

    it('should reject an IFSC with digits in the bank code', () => {
      expect(validateIFSC('HD120001234')).toBe(false);
    });

    it('should reject an IFSC that is too short', () => {
      expect(validateIFSC('HDFC000123')).toBe(false);
    });

    it('should reject an IFSC that is too long', () => {
      expect(validateIFSC('HDFC00012345')).toBe(false);
    });

    it('should reject an IFSC containing special characters', () => {
      expect(validateIFSC('HDFC0-01234')).toBe(false);
    });
  });

  describe('invalid arguments', () => {
    it('should throw for an empty string', () => {
      expect(() => validateIFSC('')).toThrow(/IFSC/);
    });

    it('should throw for a non-string input', () => {
      // @ts-expect-error testing runtime guard against non-string input
      expect(() => validateIFSC(null)).toThrow(/IFSC/);
    });
  });
});

File: packages/i18nify-go/modules/banking/validate_ifsc_test.go
package banking

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// ── ValidateIFSC ───────────────────────────────────────────────────────────────

func TestValidateIFSC(t *testing.T) {
	tests := []struct {
		name    string
		ifsc    string
		want    bool
		wantErr bool
	}{
		{name: "valid IFSC HDFC0001234", ifsc: "HDFC0001234", want: true},
		{name: "valid IFSC SBIN0000001", ifsc: "SBIN0000001", want: true},
		{name: "branch code with letters", ifsc: "ICIC0ABC123", want: true},
		{name: "lowercase IFSC is accepted", ifsc: "hdfc0001234", want: true},
		{name: "surrounding whitespace is trimmed", ifsc: "  HDFC0001234  ", want: true},

		{name: "5th char not 0", ifsc: "HDFC1001234", want: false},
		{name: "digit in bank code", ifsc: "HD120001234", want: false},
		{name: "too short", ifsc: "HDFC000123", want: false},
		{name: "too long", ifsc: "HDFC00012345", want: false},
		{name: "contains special character", ifsc: "HDFC0-01234", want: false},

		{name: "empty string returns error", ifsc: "", wantErr: true},
		{name: "whitespace-only returns error", ifsc: "   ", wantErr: true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ValidateIFSC(tt.ifsc)
			if tt.wantErr {
				assert.Error(t, err)
				return
			}
			assert.NoError(t, err)
			assert.Equal(t, tt.want, got)
		})
	}
}
